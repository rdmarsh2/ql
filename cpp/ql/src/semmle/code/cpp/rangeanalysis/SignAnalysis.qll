/**
 * Provides sign analysis to determine whether expression are always positive
 * or negative.
 *
 * The analysis is implemented as an abstract interpretation over the
 * three-valued domain `{negative, zero, positive}`.
 */
import cpp
private import semmle.code.cpp.ir.IR
private import semmle.code.cpp.controlflow.IRGuards

private newtype TSign = TNeg() or TZero() or TPos()
private class Sign extends TSign {
  string toString() {
    result = "-" and this = TNeg() or
    result = "0" and this = TZero() or
    result = "+" and this = TPos()
  }
  Sign inc() {
    this = TNeg() and result = TNeg() or
    this = TNeg() and result = TZero() or
    this = TZero() and result = TPos() or
    this = TPos() and result = TPos()
  }
  Sign dec() {
    result.inc() = this
  }
  Sign neg() {
    this = TNeg() and result = TPos() or
    this = TZero() and result = TZero() or
    this = TPos() and result = TNeg()
  }
  Sign bitnot() {
    this = TNeg() and result = TPos() or
    this = TNeg() and result = TZero() or
    this = TZero() and result = TNeg() or
    this = TPos() and result = TNeg()
  }
  Sign add(Sign s) {
    this = TZero() and result = s or
    s = TZero() and result = this or
    this = s and this = result or
    this = TPos() and s = TNeg() or
    this = TNeg() and s = TPos()
  }
  Sign mul(Sign s) {
    result = TZero() and this = TZero() or
    result = TZero() and s = TZero() or
    result = TNeg() and this = TPos() and s = TNeg() or
    result = TNeg() and this = TNeg() and s = TPos() or
    result = TPos() and this = TPos() and s = TPos() or
    result = TPos() and this = TNeg() and s = TNeg()
  }
  Sign div(Sign s) {
    result = TZero() and s = TNeg() or
    result = TZero() and s = TPos() or
    result = TNeg() and this = TPos() and s = TNeg() or
    result = TNeg() and this = TNeg() and s = TPos() or
    result = TPos() and this = TPos() and s = TPos() or
    result = TPos() and this = TNeg() and s = TNeg()
  }
  Sign rem(Sign s) {
    result = TZero() and s = TNeg() or
    result = TZero() and s = TPos() or
    result = this and s = TNeg() or
    result = this and s = TPos()
  }
  Sign bitand(Sign s) {
    result = TZero() and this = TZero() or
    result = TZero() and s = TZero() or
    result = TZero() and this = TPos() or
    result = TZero() and s = TPos() or
    result = TNeg() and this = TNeg() and s = TNeg() or
    result = TPos() and this = TNeg() and s = TPos() or
    result = TPos() and this = TPos() and s = TNeg() or
    result = TPos() and this = TPos() and s = TPos()
  }
  Sign bitor(Sign s) {
    result = TZero() and this = TZero() and s = TZero() or
    result = TNeg() and this = TNeg() or
    result = TNeg() and s = TNeg() or
    result = TPos() and this = TPos() and s = TZero() or
    result = TPos() and this = TZero() and s = TPos() or
    result = TPos() and this = TPos() and s = TPos()
  }
  Sign bitxor(Sign s) {
    result = TZero() and this = s or
    result = this and s = TZero() or
    result = s and this = TZero() or
    result = TPos() and this = TPos() and s = TPos() or
    result = TNeg() and this = TNeg() and s = TPos() or
    result = TNeg() and this = TPos() and s = TNeg() or
    result = TPos() and this = TNeg() and s = TNeg()
  }
  Sign lshift(Sign s) {
    result = TZero() and this = TZero() or
    result = this and s = TZero() or
    this != TZero() and s != TZero()
  }
  Sign rshift(Sign s) {
    result = TZero() and this = TZero() or
    result = this and s = TZero() or
    result = TNeg() and this = TNeg() or
    result != TNeg() and this = TPos() and s != TZero()
  }
  Sign urshift(Sign s) {
    result = TZero() and this = TZero() or
    result = this and s = TZero() or
    result != TZero() and this = TNeg() and s != TZero() or
    result != TNeg() and this = TPos() and s != TZero()
  }
}

private Sign certainInstructionSign(Instruction inst) {
  exists(int i | inst.(IntegerConstantInstruction).getValue().toInt() = i |
    i < 0 and result = TNeg() or
    i = 0 and result = TZero() or
    i > 0 and result = TPos()
  )
  or
  not inst instanceof IntegerConstantInstruction and
  exists(float f | f = inst.(FloatConstantInstruction).getValue().toFloat() |
    f < 0 and result = TNeg() or
    f = 0 and result = TZero() or
    f > 0 and result = TPos()
  )
}


/** Holds if the sign of `e` is too complicated to determine. */
private predicate unknownSign(Instruction i) {
  (
    i instanceof UnmodeledDefinitionInstruction or
    i instanceof UninitializedInstruction or
    i instanceof InitializeParameterInstruction or
    i instanceof BuiltInInstruction or
    i instanceof CallInstruction
  )
}

/**
 * Holds if `lowerbound` is a lower bound for `compared` at `pos`. This is restricted
 * to only include bounds for which we might determine a sign.
 */
private predicate lowerBound(Instruction lowerbound, Instruction compared, Instruction pos, boolean isStrict) {
  exists(int adjustment, IRGuardCondition comp |
    pos.getAnOperand() = compared and
  /*
   *  Java library uses guardControlsSsaRead here. I think that the phi node logic doesn't need to
   * be duplicated but the implication predicates may need to be ported
   */
   (
     isStrict = true and
     adjustment = 0
     or
     isStrict = false and
     adjustment = 1
   ) and
   comp.ensuresLt(lowerbound, compared, 0, pos.getBlock(), true)
  )
}


/**
 * Holds if `upperbound` is an upper bound for `compared` at `pos`. This is restricted
 * to only include bounds for which we might determine a sign.
 */
private predicate upperBound(Instruction upperbound, Instruction compared, Instruction pos, boolean isStrict) {
  exists(int adjustment, IRGuardCondition comp |
    pos.getAnOperand() = compared and
  /*
   * Java library uses guardControlsSsaRead here. I think that the phi node logic doesn't need to
   * be duplicated but the implication predicates may need to be ported
   */
   (
     isStrict = true and
     adjustment = 0
     or
     isStrict = false and
     adjustment = 1
   ) and
   comp.ensuresLt(compared, upperbound, 0, pos.getBlock(), true)
  )
}

/**
 * Holds if `eqbound` is an equality/inequality for `v` at `pos`. This is
 * restricted to only include bounds for which we might determine a sign. The
 * boolean `isEq` gives the polarity:
 *  - `isEq = true` : `v = eqbound`
 *  - `isEq = false` : `v != eqbound`
 */
private predicate eqBound(Instruction eqbound, Instruction compared, Instruction pos, boolean isEq) {
  exists(IRGuardCondition guard |
    pos.getAnOperand() = compared and
    guard.ensuresEq(compared, eqbound, 0, pos.getBlock(), isEq)
  )
}



/**
 * Holds if `bound` is a bound for `v` at `pos` that needs to be positive in
 * order for `v` to be positive.
 */
private predicate posBound(Instruction bound, Instruction v, Instruction pos) {
  upperBound(bound, v, pos, _) or
  eqBound(bound, v, pos, true)
}

/**
 * Holds if `bound` is a bound for `v` at `pos` that needs to be negative in
 * order for `v` to be negative.
 */
private predicate negBound(Instruction bound, Instruction v, Instruction pos) {
  lowerBound(bound, v, pos, _) or
  eqBound(bound, v, pos, true)
}

/**
 * Holds if `bound` is a bound for `v` at `pos` that can restrict whether `v`
 * can be zero.
 */
private predicate zeroBound(Instruction bound, Instruction v, Instruction pos) {
  lowerBound(bound, v, pos, _) or
  upperBound(bound, v, pos, _) or
  eqBound(bound, v, pos, _)
}

/** Holds if `bound` allows `v` to be positive at `pos`. */
private predicate posBoundOk(Instruction bound, Instruction v, Instruction pos) {
  posBound(bound, v, pos) and TPos() = instructionSign(bound)
}

/** Holds if `bound` allows `v` to be negative at `pos`. */
private predicate negBoundOk(Instruction bound, Instruction v, Instruction pos) {
  negBound(bound, v, pos) and TNeg() = instructionSign(bound)
}

/** Holds if `bound` allows `v` to be zero at `pos`. */
private predicate zeroBoundOk(Instruction bound, Instruction v, Instruction pos) {
  lowerBound(bound, v, pos, _) and TNeg() = instructionSign(bound) or
  lowerBound(bound, v, pos, false) and TZero() = instructionSign(bound) or
  upperBound(bound, v, pos, _) and TPos() = instructionSign(bound) or
  upperBound(bound, v, pos, false) and TZero() = instructionSign(bound) or
  eqBound(bound, v, pos, true) and TZero() = instructionSign(bound) or
  eqBound(bound, v, pos, false) and TZero() != instructionSign(bound)
}

private Sign binaryOpLhsSign(Instruction i) {
  result = operandSign(i, i.(BinaryInstruction).getLeftOperand())
}

private Sign binaryOpRhsSign(Instruction i) {
  result = operandSign(i, i.(BinaryInstruction).getRightOperand())
}
pragma[noinline]
private predicate binaryOpSigns(Instruction i, Sign lhs, Sign rhs) {
  lhs = binaryOpLhsSign(i) and
  rhs = binaryOpRhsSign(i)
}

/**
 * Holds if there is a bound that might restrict whether `v` has the sign `s`
 * at `pos`.
 */
private predicate hasGuard(Instruction v, Instruction pos, Sign s) {
  s = TPos() and posBound(_, v, pos) or
  s = TNeg() and negBound(_, v, pos) or
  s = TZero() and zeroBound(_, v, pos)
}

cached
private Sign operandSign(Instruction pos, Instruction operand) {
  hasGuard(operand, pos, result)
  or
  not hasGuard(operand, pos, _) and
  result = instructionSign(operand)
}

cached
private Sign instructionSign(Instruction i) {
  result = certainInstructionSign(i)
  or
  not exists(certainInstructionSign(i)) and
  (
    unknownSign(i)
    or
    exists(Instruction prior |
      prior = i.(CopyInstruction).getSourceValue()
      |
      hasGuard(prior, i, result)
      or
      not exists(Sign s | hasGuard(prior, i, s)) and
      result = instructionSign(prior)
    )
    or
    result = instructionSign(i.(BitComplementInstruction).getOperand()).bitnot()
    or
    result = instructionSign(i.(AddInstruction))
    or
    exists(Sign s1, Sign s2 |
      binaryOpSigns(i, s1, s2)
      |
      i instanceof AddInstruction and result = s1.add(s2)
      or
      i instanceof SubInstruction and result = s1.add(s2.neg())
      or
      i instanceof MulInstruction and result = s1.mul(s2)
      or
      i instanceof DivInstruction and result = s1.div(s2)
      or
      i instanceof RemInstruction and result = s1.rem(s2)
      or
      i instanceof BitAndInstruction and result = s1.bitand(s2)
      or
      i instanceof BitOrInstruction and result = s1.bitor(s2)
      or
      i instanceof BitXorInstruction and result = s1.bitxor(s2)
      or
      i instanceof ShiftLeftInstruction and result = s1.lshift(s2)
      or
      i instanceof ShiftRightInstruction and
      i.getResultType().(IntegralType).isSigned() and
      result = s1.rshift(s2)
      or
      i instanceof ShiftRightInstruction and
      not i.getResultType().(IntegralType).isSigned() and
      result = s1.urshift(s2)
    )
    or
    // use hasGuard here?
    result = operandSign(i, i.(PhiInstruction).getAnOperand())
  )
}

/** Holds if `e` can be positive and cannot be negative. */
predicate positive(Instruction i) {
  instructionSign(i) = TPos() and
  not instructionSign(i) = TNeg()
}

/** Holds if `e` can be negative and cannot be positive. */
predicate negative(Instruction i) {
  instructionSign(i) = TNeg() and
  not instructionSign(i) = TPos()
}

/** Holds if `e` is strictly positive. */
predicate strictlyPositive(Instruction i) {
  instructionSign(i) = TPos() and
  not instructionSign(i) = TNeg() and
  not instructionSign(i) = TZero()
}

/** Holds if `e` is strictly negative. */
predicate strictlyNegative(Instruction i) {
  instructionSign(i) = TNeg() and
  not instructionSign(i) = TPos() and
  not instructionSign(i) = TZero()
}
