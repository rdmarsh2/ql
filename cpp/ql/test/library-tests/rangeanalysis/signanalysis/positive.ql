import semmle.code.cpp.rangeanalysis.SignAnalysis
import semmle.code.cpp.ir.IR

from Instruction i
where positive(i)
select i, i.getAST()