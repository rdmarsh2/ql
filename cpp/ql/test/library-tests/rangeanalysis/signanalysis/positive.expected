| inline_assembly.c:10:7:10:7 | r0_7(unsigned int) = Constant[1]  | inline_assembly.c:10:7:10:7 | (unsigned int)... |
| minmax.c:16:9:16:10 | r0_3(int) = Constant[1]  | minmax.c:16:9:16:10 | 1 |
| minmax.c:16:16:16:17 | r0_6(int) = Constant[2]  | minmax.c:16:16:16:17 | 2 |
| minmax.c:16:23:16:24 | r0_9(int) = Constant[3]  | minmax.c:16:23:16:24 | 3 |
| test.c:8:19:8:19 | r2_2(int) = Constant[1]  | test.c:8:19:8:19 | 1 |
| test.c:16:20:16:20 | r2_2(int) = Constant[1]  | test.c:16:20:16:20 | 1 |
| test.c:16:25:16:26 | r2_4(int) = Constant[10]  | test.c:16:25:16:26 | 10 |
| test.c:24:5:24:11 | r2_2(int) = Constant[1]  | test.c:24:5:24:11 | ... ++ |
| test.c:25:21:25:22 | r2_7(int) = Constant[10]  | test.c:25:21:25:22 | 10 |
| test.c:33:19:33:19 | r1_4(int) = Constant[2]  | test.c:33:19:33:19 | 2 |
| test.c:33:28:33:28 | r2_8(int) = Constant[1]  | test.c:33:28:33:28 | 1 |
| test.c:42:19:42:19 | r1_4(int) = Constant[2]  | test.c:42:19:42:19 | 2 |
| test.c:42:22:42:24 | r2_8(int) = Constant[1]  | test.c:42:22:42:24 | ... ++ |
| test.c:51:17:51:17 | r1_4(int) = Constant[2]  | test.c:51:17:51:17 | 2 |
| test.c:51:21:51:21 | r1_6(int) = Constant[4]  | test.c:51:21:51:21 | 4 |
| test.c:51:30:51:30 | r2_8(int) = Constant[1]  | test.c:51:30:51:30 | 1 |
| test.c:58:11:58:11 | r0_6(int) = Constant[4]  | test.c:58:11:58:11 | 4 |
| test.c:59:13:59:13 | r2_2(int) = Constant[5]  | test.c:59:13:59:13 | 5 |
| test.c:63:10:63:10 | r4_1(int) = Constant[1]  | test.c:63:10:63:10 | 1 |
| test.c:67:24:67:25 | r2_2(int) = Constant[10]  | test.c:67:24:67:25 | 10 |
| test.c:68:15:68:15 | r3_4(int) = Constant[2]  | test.c:68:15:68:15 | 2 |
| test.c:77:13:77:13 | r2_2(int) = Constant[4]  | test.c:77:13:77:13 | 4 |
| test.c:81:13:81:13 | r4_2(int) = Constant[4]  | test.c:81:13:81:13 | 4 |
| test.c:82:14:82:14 | r5_1(int) = Constant[1]  | test.c:82:14:82:14 | 1 |
| test.c:89:11:89:11 | r0_8(int) = Constant[7]  | test.c:89:11:89:11 | 7 |
| test.c:95:10:95:10 | r5_1(int) = Constant[1]  | test.c:95:10:95:10 | 1 |
| test.c:102:6:102:8 | r2_3(int) = Constant[1]  | test.c:102:6:102:8 | ... ++ |
| test.c:104:12:104:14 | r3_4(int) = Constant[58]  | test.c:104:12:104:14 | 58 |
| test.c:107:8:107:10 | r5_3(int) = Constant[1]  | test.c:107:8:107:10 | ... ++ |
| test.c:109:14:109:16 | r6_3(int) = Constant[44]  | test.c:109:14:109:16 | 44 |
| test.c:110:14:110:14 | r7_1(int) = Constant[1]  | test.c:110:14:110:14 | 1 |
| test.c:119:10:119:12 | r0_8(unsigned long long) = Constant[1]  | test.c:119:10:119:12 | ... ++ |
| test.c:124:36:124:36 | r1_5(unsigned long long) = Constant[1]  | test.c:124:36:124:36 | (unsigned long long)... |
| test.c:127:24:127:24 | r2_6(unsigned long long) = Constant[1]  | test.c:127:24:127:24 | (unsigned long long)... |
| test.c:130:11:130:11 | r3_1(int) = Constant[1]  | test.c:130:11:130:11 | 1 |
| test.c:137:22:137:22 | r0_17(unsigned int) = Constant[1]  | test.c:137:22:137:22 | (unsigned int)... |
| test.c:138:13:138:13 | r0_23(int) = Constant[1]  | test.c:138:13:138:13 | 1 |
| test.c:161:7:161:7 | r0_7(int) = Constant[3]  | test.c:161:7:161:7 | 3 |
| test.c:161:22:161:23 | r8_2(int) = Constant[11]  | test.c:161:22:161:23 | 11 |
| test.c:166:22:166:23 | r14_2(int) = Constant[11]  | test.c:166:22:166:23 | 11 |
| test.c:171:23:171:24 | r17_2(int) = Constant[11]  | test.c:171:23:171:24 | 11 |
| test.c:176:23:176:23 | r2_2(int) = Constant[1]  | test.c:176:23:176:23 | 1 |
| test.c:200:7:200:7 | r0_9(int) = Constant[3]  | test.c:200:7:200:7 | 3 |
| test.c:200:22:200:23 | r14_2(int) = Constant[11]  | test.c:200:22:200:23 | 11 |
| test.c:200:28:200:28 | r15_0(int) = Constant[5]  | test.c:200:28:200:28 | 5 |
| test.c:200:43:200:44 | r16_2(int) = Constant[23]  | test.c:200:43:200:44 | 23 |
| test.c:204:7:204:7 | r18_1(int) = Constant[3]  | test.c:204:7:204:7 | 3 |
| test.c:204:22:204:23 | r19_2(int) = Constant[11]  | test.c:204:22:204:23 | 11 |
| test.c:204:43:204:44 | r21_2(int) = Constant[23]  | test.c:204:43:204:44 | 23 |
| test.c:208:7:208:7 | r23_1(int) = Constant[3]  | test.c:208:7:208:7 | 3 |
| test.c:208:22:208:23 | r24_2(int) = Constant[11]  | test.c:208:22:208:23 | 11 |
| test.c:208:45:208:46 | r1_2(int) = Constant[23]  | test.c:208:45:208:46 | 23 |
| test.c:212:7:212:7 | r3_1(int) = Constant[3]  | test.c:212:7:212:7 | 3 |
| test.c:212:22:212:23 | r4_2(int) = Constant[11]  | test.c:212:22:212:23 | 11 |
| test.c:216:7:216:7 | r8_1(int) = Constant[3]  | test.c:216:7:216:7 | 3 |
| test.c:216:22:216:23 | r9_2(int) = Constant[11]  | test.c:216:22:216:23 | 11 |
| test.c:228:22:228:23 | r14_2(int) = Constant[11]  | test.c:228:22:228:23 | 11 |
| test.c:228:28:228:28 | r15_0(int) = Constant[5]  | test.c:228:28:228:28 | 5 |
| test.c:228:43:228:44 | r16_2(int) = Constant[23]  | test.c:228:43:228:44 | 23 |
| test.c:232:22:232:23 | r19_2(int) = Constant[11]  | test.c:232:22:232:23 | 11 |
| test.c:232:43:232:44 | r21_2(int) = Constant[23]  | test.c:232:43:232:44 | 23 |
| test.c:236:22:236:23 | r24_2(int) = Constant[11]  | test.c:236:22:236:23 | 11 |
| test.c:236:45:236:46 | r1_2(int) = Constant[23]  | test.c:236:45:236:46 | 23 |
| test.c:240:22:240:23 | r4_2(int) = Constant[11]  | test.c:240:22:240:23 | 11 |
| test.c:244:22:244:23 | r9_2(int) = Constant[11]  | test.c:244:22:244:23 | 11 |
| test.c:256:24:256:25 | r14_2(int) = Constant[11]  | test.c:256:24:256:25 | 11 |
| test.c:256:30:256:30 | r15_0(int) = Constant[5]  | test.c:256:30:256:30 | 5 |
| test.c:256:45:256:46 | r16_2(int) = Constant[23]  | test.c:256:45:256:46 | 23 |
| test.c:260:24:260:25 | r19_2(int) = Constant[11]  | test.c:260:24:260:25 | 11 |
| test.c:260:45:260:46 | r21_2(int) = Constant[23]  | test.c:260:45:260:46 | 23 |
| test.c:264:24:264:25 | r24_2(int) = Constant[11]  | test.c:264:24:264:25 | 11 |
| test.c:264:47:264:48 | r1_2(int) = Constant[23]  | test.c:264:47:264:48 | 23 |
| test.c:268:24:268:25 | r4_2(int) = Constant[11]  | test.c:268:24:268:25 | 11 |
| test.c:272:24:272:25 | r9_2(int) = Constant[11]  | test.c:272:24:272:25 | 11 |
| test.c:284:29:284:29 | r15_0(int) = Constant[5]  | test.c:284:29:284:29 | 5 |
| test.c:284:44:284:45 | r16_2(int) = Constant[23]  | test.c:284:44:284:45 | 23 |
| test.c:288:44:288:45 | r21_2(int) = Constant[23]  | test.c:288:44:288:45 | 23 |
| test.c:292:46:292:47 | r1_2(int) = Constant[23]  | test.c:292:46:292:47 | 23 |
| test.c:312:30:312:30 | r15_0(int) = Constant[5]  | test.c:312:30:312:30 | 5 |
| test.c:312:45:312:46 | r16_2(int) = Constant[23]  | test.c:312:45:312:46 | 23 |
| test.c:316:45:316:46 | r21_2(int) = Constant[23]  | test.c:316:45:316:46 | 23 |
| test.c:320:47:320:48 | r1_2(int) = Constant[23]  | test.c:320:47:320:48 | 23 |
| test.c:342:14:342:14 | r3_3(int) = Constant[3]  | test.c:342:14:342:14 | 3 |
| test.c:343:5:343:7 | r4_2(int) = Constant[1]  | test.c:343:5:343:7 | ... ++ |
| test.c:348:14:348:14 | r7_1(int) = Constant[1]  | test.c:348:14:348:14 | 1 |
| test.c:357:12:357:14 | r0_22(unsigned int) = Constant[100]  | test.c:357:12:357:14 | (unsigned int)... |
| test.c:357:22:357:23 | r22_0(unsigned int) = Constant[10]  | test.c:357:22:357:23 | (unsigned int)... |
| test.c:358:13:358:15 | r17_7(unsigned int) = Constant[100]  | test.c:358:13:358:15 | (unsigned int)... |
| test.c:358:19:358:20 | r24_0(unsigned int) = Constant[10]  | test.c:358:19:358:20 | (unsigned int)... |
| test.c:365:11:365:13 | r23_25(unsigned int) = Constant[300]  | test.c:365:11:365:13 | (unsigned int)... |
| test.c:366:15:366:15 | r38_0(unsigned int) = Constant[5]  | test.c:366:15:366:15 | (unsigned int)... |
| test.c:366:15:366:15 | r40_0(unsigned int) = Constant[5]  | test.c:366:15:366:15 | (unsigned int)... |
| test.c:367:15:367:17 | r48_0(unsigned int) = Constant[500]  | test.c:367:15:367:17 | (unsigned int)... |
| test.c:367:15:367:17 | r50_0(unsigned int) = Constant[500]  | test.c:367:15:367:17 | (unsigned int)... |
| test.c:368:13:368:13 | r27_0(unsigned int) = Constant[1]  | test.c:368:13:368:13 | (unsigned int)... |
| test.c:368:13:368:13 | r54_0(unsigned int) = Constant[1]  | test.c:368:13:368:13 | (unsigned int)... |
| test.c:368:19:368:21 | r55_0(unsigned int) = Constant[500]  | test.c:368:19:368:21 | (unsigned int)... |
| test.c:369:29:369:29 | r4_0(unsigned int) = Constant[1]  | test.c:369:29:369:29 | (unsigned int)... |
| test.c:369:29:369:29 | r41_0(unsigned int) = Constant[1]  | test.c:369:29:369:29 | (unsigned int)... |
| test.c:369:36:369:36 | r5_0(int) = Constant[5]  | test.c:369:36:369:36 | 5 |
| test.c:370:29:370:29 | r9_0(unsigned int) = Constant[1]  | test.c:370:29:370:29 | (unsigned int)... |
| test.c:370:29:370:29 | r41_0(unsigned int) = Constant[1]  | test.c:370:29:370:29 | (unsigned int)... |
| test.c:370:36:370:38 | r10_0(int) = Constant[500]  | test.c:370:36:370:38 | 500 |
| test.c:371:30:371:30 | r14_0(unsigned int) = Constant[1]  | test.c:371:30:371:30 | (unsigned int)... |
| test.c:371:30:371:30 | r41_0(unsigned int) = Constant[1]  | test.c:371:30:371:30 | (unsigned int)... |
| test.c:371:37:371:39 | r15_0(int) = Constant[500]  | test.c:371:37:371:39 | 500 |
| test.c:379:12:379:14 | r0_16(unsigned int) = Constant[100]  | test.c:379:12:379:14 | (unsigned int)... |
| test.c:379:22:379:24 | r5_0(unsigned int) = Constant[110]  | test.c:379:22:379:24 | (unsigned int)... |
| test.c:380:13:380:15 | r1_7(unsigned int) = Constant[100]  | test.c:380:13:380:15 | (unsigned int)... |
| test.c:380:19:380:21 | r7_0(unsigned int) = Constant[110]  | test.c:380:19:380:21 | (unsigned int)... |
| test.c:381:8:381:11 | r6_5(unsigned int) = Constant[1000]  | test.c:381:8:381:11 | (unsigned int)... |
| test.c:382:8:382:11 | r6_8(unsigned int) = Constant[1000]  | test.c:382:8:382:11 | (unsigned int)... |
| test.c:383:8:383:11 | r6_11(unsigned int) = Constant[1000]  | test.c:383:8:383:11 | (unsigned int)... |
| test.c:384:12:384:14 | r6_16(unsigned int) = Constant[300]  | test.c:384:12:384:14 | (unsigned int)... |
| test.c:385:13:385:15 | r11_0(unsigned int) = Constant[300]  | test.c:385:13:385:15 | (unsigned int)... |
| test.c:385:13:385:15 | r18_0(unsigned int) = Constant[300]  | test.c:385:13:385:15 | (unsigned int)... |
| test.c:385:21:385:21 | r19_0(unsigned int) = Constant[5]  | test.c:385:21:385:21 | (unsigned int)... |
| test.c:386:13:386:15 | r11_0(unsigned int) = Constant[200]  | test.c:386:13:386:15 | (unsigned int)... |
| test.c:386:13:386:15 | r24_0(unsigned int) = Constant[200]  | test.c:386:13:386:15 | (unsigned int)... |
| test.c:386:21:386:21 | r25_0(unsigned int) = Constant[5]  | test.c:386:21:386:21 | (unsigned int)... |
| test.c:387:29:387:31 | r28_0(unsigned int) = Constant[200]  | test.c:387:29:387:31 | (unsigned int)... |
| test.c:387:29:387:31 | r30_0(unsigned int) = Constant[200]  | test.c:387:29:387:31 | (unsigned int)... |
| test.c:387:38:387:38 | r31_0(int) = Constant[5]  | test.c:387:38:387:38 | 5 |
| test.c:394:24:394:26 | r0_7(unsigned int) = Constant[100]  | test.c:394:24:394:26 | (unsigned int)... |
| test.c:394:34:394:36 | r2_0(unsigned int) = Constant[100]  | test.c:394:34:394:36 | (unsigned int)... |
| test.c:397:9:397:11 | r3_10(unsigned int) = Constant[1]  | test.c:397:9:397:11 | ++ ... |
| test.c:398:9:398:11 | r3_19(unsigned int) = Constant[1]  | test.c:398:9:398:11 | ... ++ |
| test.c:398:19:398:19 | r3_22(unsigned int) = Constant[3]  | test.c:398:19:398:19 | (unsigned int)... |
| test.cpp:11:13:11:13 | r1_2(int) = Constant[3]  | test.cpp:11:13:11:13 | 3 |
| test.cpp:33:12:33:12 | r13_3(int) = Constant[1]  | test.cpp:33:12:33:12 | 1 |
| test.cpp:34:9:34:9 | r14_0(int) = Constant[1]  | test.cpp:34:9:34:9 | 1 |
| test.cpp:39:12:39:14 | r17_3(int) = Constant[128]  | test.cpp:39:12:39:14 | 128 |
| test.cpp:40:9:40:11 | r18_0(int) = Constant[128]  | test.cpp:40:9:40:11 | 128 |
| test.cpp:45:12:45:15 | r21_3(int) = Constant[1024]  | test.cpp:45:12:45:15 | 1024 |
| test.cpp:46:9:46:12 | r22_0(int) = Constant[1024]  | test.cpp:46:9:46:12 | 1024 |
| test.cpp:69:10:69:21 | r9_1(bool) = Constant[1]  | test.cpp:69:10:69:21 | ... \|\| ... |
